File: cpp-tsdb/test/alert/manager_test.cpp
#include <gtest/gtest.h>
#include <chrono>
#include <filesystem>
#include <thread>
#include "tsdb/alert/manager.hpp"
#include "tsdb/promql/engine.hpp"
#include "tsdb/storage/storage.hpp"
#include "../test_utils.hpp"

namespace tsdb::test {

class AlertManagerTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Create temporary directory for tests
        test_path_ = std::filesystem::temp_directory_path() / "tsdb_alert_test";
        std::filesystem::create_directories(test_path_);
        
        // Initialize components
        StorageConfig config{
            .data_path = test_path_.string(),
            .max_block_size = 1024 * 1024
        };
        storage_ = std::make_unique<Storage>(config);
        engine_ = std::make_unique<PromQLEngine>(storage_.get());
        manager_ = std::make_unique<AlertManager>(engine_.get());

        // Setup test data
        setupTestData();
    }

    void TearDown() override {
        manager_.reset();
        engine_.reset();
        storage_.reset();
        std::filesystem::remove_all(test_path_);
    }

    void setupTestData() {
        auto series = TestUtils::generateTestSeries({
            {"__name__", "error_rate"},
            {"service", "api"}
        }, 100);
        storage_->writeSeries(series);
    }

    std::filesystem::path test_path_;
    std::unique_ptr<Storage> storage_;
    std::unique_ptr<PromQLEngine> engine_;
    std::unique_ptr<AlertManager> manager_;
};

TEST_F(AlertManagerTest, BasicRules) {
    // Test simple threshold rule
    {
        Rule rule{
            .name = "high_error_rate",
            .query = "error_rate > 0.5",
            .duration = std::chrono::minutes(5),
            .labels = {
                {"severity", "critical"}
            },
            .annotations = {
                {"summary", "High error rate detected"}
            }
        };

        EXPECT_NO_THROW(manager_->addRule(rule));

        // Run evaluation
        auto alerts = manager_->evaluate();
        ASSERT_TRUE(alerts);
        EXPECT_FALSE(alerts->empty());
    }

    // Test rate rule
    {
        Rule rule{
            .name = "error_spike",
            .query = "rate(error_rate[5m]) > 0.1",
            .duration = std::chrono::minutes(5),
            .labels = {
                {"severity", "warning"}
            },
            .annotations = {
                {"summary", "Error rate is increasing rapidly"}
            }
        };

        EXPECT_NO_THROW(manager_->addRule(rule));

        // Run evaluation
        auto alerts = manager_->evaluate();
        ASSERT_TRUE(alerts);
        EXPECT_FALSE(alerts->empty());
    }
}

TEST_F(AlertManagerTest, StatePersistence) {
    Rule rule{
        .name = "test_alert",
        .query = "error_rate > 0.5",
        .duration = std::chrono::minutes(5),
        .labels = {
            {"severity", "warning"}
        }
    };

    EXPECT_NO_THROW(manager_->addRule(rule));

    // Test alert state transitions
    auto alerts1 = manager_->evaluate();
    ASSERT_TRUE(alerts1);
    
    auto alerts2 = manager_->evaluate();
    ASSERT_TRUE(alerts2);

    // Check if state is maintained
    EXPECT_EQ(alerts1->size(), alerts2->size());
    for (size_t i = 0; i < alerts1->size(); ++i) {
        EXPECT_EQ((*alerts1)[i].state, (*alerts2)[i].state);
    }
}

TEST_F(AlertManagerTest, Notifications) {
    // Create notification channel
    auto notifications = std::make_shared<AlertChannel>();
    manager_ = std::make_unique<AlertManager>(engine_.get(), notifications);

    Rule rule{
        .name = "test_notification",
        .query = "error_rate > 0.5",
        .duration = std::chrono::seconds(1),
        .labels = {
            {"severity", "critical"}
        }
    };

    EXPECT_NO_THROW(manager_->addRule(rule));

    // Run evaluation
    auto alerts = manager_->evaluate();
    ASSERT_TRUE(alerts);

    // Check for notification
    Alert alert;
    EXPECT_TRUE(notifications->try_dequeue(alert));
    EXPECT_EQ(alert.name, "test_notification");
}

TEST_F(AlertManagerTest, Concurrency) {
    // Add multiple rules
    std::vector<Rule> rules = {
        {
            .name = "rule1",
            .query = "metric1 > 0.5",
            .duration = std::chrono::minutes(5)
        },
        {
            .name = "rule2",
            .query = "metric2 > 0.5",
            .duration = std::chrono::minutes(5)
        },
        {
            .name = "rule3",
            .query = "metric3 > 0.5",
            .duration = std::chrono::minutes(5)
        }
    };

    for (const auto& rule : rules) {
        EXPECT_NO_THROW(manager_->addRule(rule));
    }

    // Test concurrent evaluations
    std::vector<std::thread> threads;
    std::atomic<int> errors{0};

    for (int i = 0; i < 10; ++i) {
        threads.emplace_back([this, &errors]() {
            try {
                auto alerts = manager_->evaluate();
                if (!alerts) {
                    errors++;
                }
            } catch (...) {
                errors++;
            }
        });
    }

    for (auto& thread : threads) {
        thread.join();
    }

    EXPECT_EQ(errors, 0);
}

} // namespace tsdb::test 

File: go-tsdb/internal/alert/manager_test.go
package alert

import (
	"context"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/vobbilis/go-tsdb/internal/promql"
	"github.com/vobbilis/go-tsdb/internal/storage"
	"github.com/vobbilis/go-tsdb/test/testutil"
)

func TestAlertBasicRules(t *testing.T) {
	// Setup storage and PromQL engine
	store, err := storage.NewStorage(storage.StorageConfig{
		DataPath:     t.TempDir(),
		MaxBlockSize: 1024 * 1024,
	})
	require.NoError(t, err)
	defer store.Close()

	engine := promql.NewEngine(store)
	manager := NewAlertManager(engine)

	// Setup test data
	series := testutil.GenerateTestSeries(map[string]string{
		"__name__": "error_rate",
		"service":  "api",
	}, 100)
	err = store.WriteSeries(series)
	require.NoError(t, err)

	// Test simple threshold rule
	t.Run("ThresholdRule", func(t *testing.T) {
		rule := &Rule{
			Name:     "high_error_rate",
			Query:    `error_rate > 0.5`,
			Duration: 5 * time.Minute,
			Labels: map[string]string{
				"severity": "critical",
			},
			Annotations: map[string]string{
				"summary": "High error rate detected",
			},
		}

		err := manager.AddRule(rule)
		assert.NoError(t, err)

		// Run evaluation
		ctx := context.Background()
		alerts, err := manager.Evaluate(ctx)
		assert.NoError(t, err)
		assert.NotEmpty(t, alerts)
	})

	// Test rate rule
	t.Run("RateRule", func(t *testing.T) {
		rule := &Rule{
			Name:     "error_spike",
			Query:    `rate(error_rate[5m]) > 0.1`,
			Duration: 5 * time.Minute,
			Labels: map[string]string{
				"severity": "warning",
			},
			Annotations: map[string]string{
				"summary": "Error rate is increasing rapidly",
			},
		}

		err := manager.AddRule(rule)
		assert.NoError(t, err)

		// Run evaluation
		ctx := context.Background()
		alerts, err := manager.Evaluate(ctx)
		assert.NoError(t, err)
		assert.NotEmpty(t, alerts)
	})
}

func TestAlertStatePersistence(t *testing.T) {
	store, err := storage.NewStorage(storage.StorageConfig{
		DataPath:     t.TempDir(),
		MaxBlockSize: 1024 * 1024,
	})
	require.NoError(t, err)
	defer store.Close()

	engine := promql.NewEngine(store)
	manager := NewAlertManager(engine)

	// Add test rule
	rule := &Rule{
		Name:     "test_alert",
		Query:    `error_rate > 0.5`,
		Duration: 5 * time.Minute,
		Labels: map[string]string{
			"severity": "warning",
		},
	}

	err = manager.AddRule(rule)
	require.NoError(t, err)

	// Test alert state transitions
	t.Run("StateTransitions", func(t *testing.T) {
		ctx := context.Background()

		// First evaluation
		alerts1, err := manager.Evaluate(ctx)
		assert.NoError(t, err)

		// Second evaluation
		alerts2, err := manager.Evaluate(ctx)
		assert.NoError(t, err)

		// Check if state is maintained
		assert.Equal(t, len(alerts1), len(alerts2))
		for i := range alerts1 {
			assert.Equal(t, alerts1[i].State, alerts2[i].State)
		}
	})
}

func TestAlertNotifications(t *testing.T) {
	store, err := storage.NewStorage(storage.StorageConfig{
		DataPath:     t.TempDir(),
		MaxBlockSize: 1024 * 1024,
	})
	require.NoError(t, err)
	defer store.Close()

	engine := promql.NewEngine(store)

	// Create notification channel
	notifications := make(chan *Alert, 100)
	manager := NewAlertManager(engine, WithNotificationChannel(notifications))

	// Add test rule
	rule := &Rule{
		Name:     "test_notification",
		Query:    `error_rate > 0.5`,
		Duration: 1 * time.Second,
		Labels: map[string]string{
			"severity": "critical",
		},
	}

	err = manager.AddRule(rule)
	require.NoError(t, err)

	// Test notification delivery
	t.Run("NotificationDelivery", func(t *testing.T) {
		ctx := context.Background()

		// Run evaluation
		_, err := manager.Evaluate(ctx)
		assert.NoError(t, err)

		// Check for notification
		select {
		case alert := <-notifications:
			assert.NotNil(t, alert)
			assert.Equal(t, "test_notification", alert.Name)
		case <-time.After(5 * time.Second):
			t.Fatal("Notification timeout")
		}
	})
}

func TestAlertConcurrency(t *testing.T) {
	store, err := storage.NewStorage(storage.StorageConfig{
		DataPath:     t.TempDir(),
		MaxBlockSize: 1024 * 1024,
	})
	require.NoError(t, err)
	defer store.Close()

	engine := promql.NewEngine(store)
	manager := NewAlertManager(engine)

	// Add multiple rules
	rules := []*Rule{
		{
			Name:     "rule1",
			Query:    `metric1 > 0.5`,
			Duration: 5 * time.Minute,
		},
		{
			Name:     "rule2",
			Query:    `metric2 > 0.5`,
			Duration: 5 * time.Minute,
		},
		{
			Name:     "rule3",
			Query:    `metric3 > 0.5`,
			Duration: 5 * time.Minute,
		},
	}

	for _, rule := range rules {
		err := manager.AddRule(rule)
		require.NoError(t, err)
	}

	// Test concurrent evaluations
	t.Run("ConcurrentEvaluations", func(t *testing.T) {
		ctx := context.Background()
		done := make(chan error, 10)

		for i := 0; i < 10; i++ {
			go func() {
				_, err := manager.Evaluate(ctx)
				done <- err
			}()
		}

		for i := 0; i < 10; i++ {
			err := <-done
			assert.NoError(t, err)
		}
	})
}
